scanpy/_utils/__init__.py

```diff
@@ -560,6 +569,160 @@ def _elem_mul_dask(x: DaskArray, y: DaskArray) -> DaskArray:
     return da.map_blocks(elem_mul, x, y)
 
 
+Scaling_T = TypeVar("Scaling_T", DaskArray, np.ndarray)
+
+
+def broadcast_axis(divisor: Scaling_T, axis: Literal[0, 1]) -> Scaling_T:
+    divisor = np.ravel(divisor)
+    if axis:
+        return divisor[None, :]
+    return divisor[:, None]
+
+
+@singledispatch
+def axis_scale(
+    X: sparse.spmatrix,
+    scaling_array,
+    axis: Literal[0, 1],
+    *,
+    out: sparse.spmatrix | None = None,
+) -> sparse.spmatrix:
+    if X is out:
+        if axis == 0:
+            sparsefuncs.inplace_row_scale(X, np.ravel(scaling_array))
+            return X
+        sparsefuncs.inplace_column_scale(X, np.ravel(scaling_array))
+        return X
+
+    row_scale = axis == 0
+    column_scale = axis == 1
+    # Same as `inplace_{column,row}_scale`, but not in place
+    if X.format == "csr":
+        if row_scale:
+            new_data = X.data * np.repeat(scaling_array, np.diff(X.indptr))
+        elif column_scale:
+            new_data = X.data * scaling_array.take(X.indices, mode="clip")
+        return sparse.csr_matrix((new_data, X.indices, X.indptr), shape=X.shape)
+    return axis_scale(X.T, scaling_array, (axis + 1) % 2).T
+
+
+@axis_scale.register(np.ndarray)
+def _(
+    X: np.ndarray,
+    scaling_array: np.ndarray,
+    axis: Literal[0, 1],
+    *,
+    out: np.ndarray | None = None,
+) -> np.ndarray:
+    scaling_array = broadcast_axis(scaling_array, axis)
+    scaled_X = np.multiply(X, scaling_array, out=out)
+    return scaled_X
+
+
+@axis_scale.register(DaskArray)
+def _(
+    X: DaskArray,
+    scaling_array: Scaling_T,
+    axis: Literal[0, 1],
+    *,
+    out: None = None,
+) -> DaskArray:
+    if out is not None:
+        raise TypeError(
+            "`out` is not `None`. Do not do in-place modifications on dask arrays."
+        )
+
+    import dask.array as da
+
+    scaling_array = broadcast_axis(scaling_array, axis)
+    row_scale = axis == 0
+    column_scale = axis == 1
+
+    if isinstance(scaling_array, DaskArray):
+        warnings.warn("Rechunking scaling_array in user operation", UserWarning)
```

Comment:
I think rechunking is a non-trivial operation but this also is not a public API.  So maybe it should be removed then?

Reply:
Sure, but this is just warning whenever a dask array passes through, and isn't specific to the case where rechecking is necessary.

I believe rechecking generally shouldn't be necessary here, since the chunks should be aligned already (though it would be nice to enforce this).

```python
row_chunks = np.random.randint(10, 20, size=10)
X = da.random.poisson(size=(row_chunks.sum(), 100), chunks=(tuple(row_chunks), 50))

assert X.sum(axis=1).chunks[0] == X.chunks[0]
```